unified_planning/engines/oversubscription_planner.py

```diff
@@ -0,0 +1,124 @@
+        supported_kind = ProblemKind()
+        supported_kind.set_problem_class("ACTION_BASED")
+        supported_kind.set_typing("FLAT_TYPING")
+        supported_kind.set_typing("HIERARCHICAL_TYPING")
+        supported_kind.set_numbers("CONTINUOUS_NUMBERS")
+        supported_kind.set_numbers("DISCRETE_NUMBERS")
+        supported_kind.set_fluents_type("NUMERIC_FLUENTS")
+        supported_kind.set_fluents_type("OBJECT_FLUENTS")
+        supported_kind.set_conditions_kind("NEGATIVE_CONDITIONS")
+        supported_kind.set_conditions_kind("DISJUNCTIVE_CONDITIONS")
+        supported_kind.set_conditions_kind("EQUALITY")
+        supported_kind.set_conditions_kind("EXISTENTIAL_CONDITIONS")
+        supported_kind.set_conditions_kind("UNIVERSAL_CONDITIONS")
+        supported_kind.set_effects_kind("CONDITIONAL_EFFECTS")
+        supported_kind.set_effects_kind("INCREASE_EFFECTS")
+        supported_kind.set_effects_kind("DECREASE_EFFECTS")
+        supported_kind.set_time("CONTINUOUS_TIME")
+        supported_kind.set_time("DISCRETE_TIME")
+        supported_kind.set_time("INTERMEDIATE_CONDITIONS_AND_EFFECTS")
+        supported_kind.set_time("TIMED_EFFECT")
+        supported_kind.set_time("TIMED_GOALS")
+        supported_kind.set_time("DURATION_INEQUALITIES")
+        supported_kind.set_expression_duration("STATIC_FLUENTS_IN_DURATION")
+        supported_kind.set_expression_duration("FLUENTS_IN_DURATION")
+        supported_kind.set_simulated_entities("SIMULATED_EFFECTS")
+        final_supported_kind = supported_kind.intersection(engine.supported_kind())
+        final_supported_kind.set_quality_metrics("OVERSUBSCRIPTION")
+        return final_supported_kind
+
+    @staticmethod
+    def _supports(problem_kind: "ProblemKind", engine: Type[Engine]) -> bool:
+        return problem_kind <= OversubscriptionPlanner._supported_kind(engine)
+
+    def _solve(
+        self,
+        problem: "up.model.AbstractProblem",
+        callback: Optional[
+            Callable[["up.engines.results.PlanGenerationResult"], None]
+        ] = None,
+        timeout: Optional[float] = None,
+        output_stream: Optional[IO[str]] = None,
+    ) -> "up.engines.results.PlanGenerationResult":
+        assert isinstance(problem, up.model.Problem)
+        assert isinstance(self.engine, mixins.OneshotPlannerMixin)
+        if len(problem.quality_metrics) == 0:
+            goals: List[Tuple["up.model.FNode", Union[Fraction, int]]] = []
+        else:
+            assert len(problem.quality_metrics) == 1
+            qm = problem.quality_metrics[0]
+            assert isinstance(qm, up.model.metrics.Oversubscription)
+            goals = list(qm.goals.items())
+        q: PriorityQueue = PriorityQueue()
+        for l in powerset(goals):
+            cost: Union[Fraction, int] = 0
+            sg = []
+            for g, c in l:
+                cost += c
+                sg.append(g)
+            q.put((-1 * cost, sg))
```

Comment:
Using tuples (cost, goal) is clever, but beware that it might happen that two goals are compared to break ties (it **should** work, but please check).